owlkettle/gtk.nim

```diff
@@ -365,7 +366,42 @@ proc g_value_unset*(value: ptr GValue)
 proc g_idle_add_full*(priority: cint, fn: GSourceFunc, data: pointer, notify: GDestroyNotify): cuint
 proc g_timeout_add_full*(priority: cint, interval: cuint, fn: GSourceFunc, data: pointer, notify: GDestroyNotify): cuint
 proc g_strdup*(str: cstring): pointer
+proc g_free*(str: pointer)
+{.pop.}
+
+type OwnedGtkString* = distinct cstring
+  # Forward declarations are needed as otherwise the line `cStr.cstring.isNil` 
+  # in the =destroy hook causes nim to use cstring's hooks implicitly, which 
+  # we do not want to ensure memory safe use of OwnedGtkString
+proc `=copy`*(dest: var OwnedGtkString, source: OwnedGtkString)
+proc `=sink`*(dest: var OwnedGtkString; source: OwnedGtkString)
+
+crossVersionDestructor(cstr, OwnedGtkString):
+  echo "Destroy owned gtk string"
+  if cStr.cstring.isNil():
+    return
+  
+  g_free(pointer(cStr))
+
+proc `=copy`*(dest: var OwnedGtkString, source: OwnedGtkString) =
+  echo "Copy owned string"
+  let areSameObject = pointer(source) == pointer(dest)
+  if areSameObject:
+    return
+  
+  `=destroy`(dest)
+  wasMoved(dest)
+  cstring(dest) = cast[cstring](g_strdup(cstring(source)))
+
+proc `=sink`*(dest: var OwnedGtkString; source: OwnedGtkString) =
+  echo "Sink owned string"
+  `=destroy`(dest)
+  wasMoved(dest)
+  dest = source
```

Comment:
```suggestion
  pointer(dest) = pointer(source)
```

Reply:
We want to move here, not `=copy`.

Reply:
Good point, my lack of being used to deal with hooks and thinking their effects through is showing =/, done

Reply:
Oh, sorry, you might need to use `cstring` instead of pointer.

Reply:
Huh, that no longer compiles because it really doesn't like assigning to pointer(dest)

Reply:
That compiles, whether it actually is functional though I still can't evaluate as I can't even imagine what the code would need to look like that'd show me that it works.